src/libraries/Microsoft.Extensions.Diagnostics/src/Metrics/ListenerSubscription.cs

```diff
@@ -0,0 +1,324 @@
+
+                if (status.Published)
+                {
+                    status.Published = false;
+                    if (status.ListenerEnabled)
+                    {
+                        _metricsListener.MeasurementsCompleted(instrument, status.State);
+                    }
+                }
+            }
+        }
+
+        internal void UpdateRules(IList<InstrumentRule> rules)
+        {
+            lock (_instruments)
+            {
+                if (_disposed)
+                {
+                    return;
+                }
+
+                _rules = rules;
+
+                foreach (var instrumentPair in _instruments)
+                {
+                    RefreshInstrument(instrumentPair.Key, instrumentPair.Value);
+                }
+            }
+        }
+
+        // Called under _instrument lock
+        private void RefreshInstrument(Instrument instrument, InstrumentStatus status)
+        {
+            var alreadyEnabled = status.Enabled;
+            var enable = false;
+            var rule = GetMostSpecificRule(instrument);
+            if (rule != null)
+            {
+                enable = rule.Enable;
+            }
+
+            if (!enable && alreadyEnabled)
+            {
+                status.Enabled = false;
+                _meterListener.DisableMeasurementEvents(instrument);
+            }
+            else if (enable && !alreadyEnabled)
+            {
+                // The first time we enable an instrument, we need to call InstrumentPublished.
+                if (!status.Published)
+                {
+                    // However, a listener might decline to enable the instrument, remember that.
+                    status.Published = true;
+                    status.ListenerEnabled = _metricsListener.InstrumentPublished(instrument, out var state);
+                    status.State = state;
+                }
+
+                if (status.ListenerEnabled)
+                {
+                    _meterListener.EnableMeasurementEvents(instrument, status.State);
+                    status.Enabled = true;
+                }
+            }
+        }
+
+        private InstrumentRule? GetMostSpecificRule(Instrument instrument)
+        {
+            InstrumentRule? best = null;
+            foreach (var rule in _rules)
+            {
+                if (RuleMatches(rule, instrument, _metricsListener.Name)
+                    && IsMoreSpecific(rule, best, isLocalScope: instrument.Meter.Scope != null))
+                {
+                    best = rule;
+                }
+            }
+
+            return best;
+        }
+
+        // internal for testing
+        internal static bool RuleMatches(InstrumentRule rule, Instrument instrument, string listenerName)
+        {
+            // Exact match or empty
+            if (!string.IsNullOrEmpty(rule.ListenerName)
+                && !string.Equals(rule.ListenerName, listenerName, StringComparison.OrdinalIgnoreCase))
```

Comment:
>!string.Equals(rule.ListenerName, listenerName, StringComparison.OrdinalIgnoreCase) [](http://example.com/codeflow?start=19&length=83)

nit: this is ok but you can write it short as `rule.ListenerName != listenerName`

Reply:
I'd rather be explicit, I can never remember what the default comparer does.

Reply:
fine with me. I was pointing better for readability and less writing :-)